File: graph/dfs.c
// program to represent graph in memory and to implement DFS traversal technique on it
// 1 adjancey matrix reps.
// 2 linked list reps.
// linked array combine reps.

#include<stdio.h>
#include<stdlib.h> // for malloc
#define max 100

typedef struct node {

	int dest;
	struct node *next;
} graphtype;

//global variable easy to access and modify
graphtype *vertice[max];

void dfs(int i,int visited[])
{
	graphtype *ptr;

	if(!visited[i])
	{
		visited[i]=1;
		printf("%d ",i);
	}

	ptr = vertice[i];

	while(ptr!=NULL)
	{
		int j;
		j=ptr->dest;

		if(!visited[j])
		{
			//printf("node processed is %d to %d\n",i+1,j+1);
			dfs(j,visited);		// recursive call
		}
		ptr=ptr->next;
	}
}

void create_edge(int s,int d)
{
	graphtype *ptr = (graphtype*)malloc(sizeof(graphtype));
	ptr->dest=d;
	ptr->next = vertice[s];
	vertice[s]=ptr;
}

void input_graph(int total_edge)
{
	int i=1;
	while(total_edge--)
	{
		int s,d;
		//printf("enter source and destination of edege %d: ",i++);
		scanf("%d%d",&s,&d);

		create_edge(s,d);
	}
}

void display_graph(int v)
{
	int e=1;

	for(int i=0; i<v; i++)
	{
		graphtype *ptr = vertice[i];
		printf("\nedge %d: ",i);
		while(ptr!=NULL)
		{
			printf("->%d",ptr->dest);
			ptr=ptr->next; e++;
		}
	}
}

int main()
{
	int v,e;

	//printf("enter total number of vertices and edges: ");
	scanf("%d%d",&v,&e);

	input_graph(e);			//O(e)
	display_graph(v);		//O(v+e)

	for(int i=0; i<v; i++)
	{
		int visited[max]={0};
		printf("\n");
		dfs(i,visited);
	}
}

File: graph/dfs.py
#python program to reprsent graph in memory

class Graph :
	"""blue print for graph data structure"""
	def __init__(self,dest = 0, next = None):
		self.dest = dest
		self.next = next

def dfs(vertices,visited,source):

	if(not visited[source]):
		print("node processed is {}".format(source))
		visited[source] = 1

	ptr = vertices[source]

	while(ptr):

		if(not visited[ptr.dest]):
			dfs(vertices,visited,ptr.dest)
		ptr = ptr.next


def input_graph(vertices,e):

	while e>0:
		e-=1
		print("enter source and destination :",end = " ")
		s,d = input().split(" ")
		s,d = int(s),int(d)

		ptr = Graph(d,vertices[s])
		vertices[s] = ptr

def display_graph(vertices):

	for ptr in vertices:

		while(ptr):
			print("[{}]-->".format(ptr.dest),end = " ")
			ptr = ptr.next
		print("NULL")

print("enter total number of vertices and edges ")

v,e = input().split(" ")
v,e = int(v),int(e)
vertices = [None] * v
visited = [0] * v

input_graph(vertices,e)
display_graph(vertices)
dfs(vertices,visited,0)